src/mlpack/methods/kde/kde_rules_impl.hpp

```diff
@@ -108,20 +122,80 @@ Score(const size_t queryIndex, TreeType& referenceNode)
 
     // Calculate kernel value based on reference node centroid.
     if (tree::TreeTraits<TreeType>::FirstPointIsCentroid)
-    {
       kernelValue = EvaluateKernel(queryIndex, referenceNode.Point(0));
-    }
     else
-    {
-      kde::KDEStat& referenceStat = referenceNode.Stat();
       kernelValue = EvaluateKernel(queryPoint, referenceStat.Centroid());
-    }
 
     densities(queryIndex) += referenceNode.NumDescendants() * kernelValue;
 
     // Don't explore this tree branch.
     score = DBL_MAX;
   }
+  else if (monteCarlo &&
+           referenceNode.NumDescendants() >= MCAccessCoef * initialSampleSize &&
+           std::is_same<KernelType, kernel::GaussianKernel>::value)
+  {
+    // Calculate reference node depth.
+    if (referenceStat.Depth() == 0)
+      referenceStat.Depth() = CalculateDepth(referenceNode);
+
+    // Monte Carlo probabilistic estimation.
+    const double currentAlpha =
+        MCBeta / std::pow(2, referenceNode.Stat().Depth() + 1);
+    const boost::math::normal normalDist;
+    const double z =
+        std::abs(boost::math::quantile(normalDist, currentAlpha / 2));
+    const size_t numDesc = referenceNode.NumDescendants();
+    arma::vec sample;
+    size_t m = initialSampleSize;
+    double meanSample;
+    bool useMonteCarloPredictions = true;
+    while (m > 0)
+    {
+      const size_t oldSize = sample.size();
+      const size_t newSize = oldSize + m;
+
+      // Don't use probabilistic estimation if this is gonna take a close
+      // amount of computation to the exact calculation.
+      if (newSize >= MCBreakCoef * numDesc)
+      {
+        useMonteCarloPredictions = false;
+        break;
+      }
+
+      // Increase the sample size.
+      sample.resize(newSize);
+      for (size_t i = 0; i < m; ++i)
+      {
+        // Sample and evaluate random points from the reference node.
+        const size_t randomPoint = math::RandInt(0, numDesc);
+        sample(oldSize + i) =
+            EvaluateKernel(queryIndex, referenceNode.Descendant(randomPoint));
+      }
```

Comment:
Hmm, I see an acceleration opportunity here, but I don't know how much speedup it would give or if it is worth the amount of work I think it might be.  We're computing the kernel evaluations for `m` points at once, and `m` could be of reasonable size (like 20 or more).  Specifically, `EvaluateKernel` sequentially calls `kernel.Evaluate(metric.Evaluate(query, reference))`, and at least in the `EuclideanDistance` case, `metric.Evaluate(query, reference)` boils down to `arma::norm(query - reference, 2)`.  But that's for just a single point; it may be possible to express the Euclidean distance calculation between *many* query and reference points as a big linear algebra expression and thus Armadillo could make better use of LAPACK or BLAS internally to do that: typically, larger matrix multiplications can be quicker than a series of small vector operations.  (Every situation is different; that may not always apply.)

Anyway, think about it, and let me know what you think.  It may be a lot of work---I haven't fully thought through all the changes.  So if you don't think that you want to do it, it's perfectly okay; I just wanted to suggest the idea.

Reply:
I'm thinking about how to do it, but in any case, wouldn't it also be nice to modify the implementation of ```src/mlpack/core/metrics/lmetric_impl.hpp ```?
From

```
typename VecTypeA::elem_type LMetric<2, true>::Evaluate(
    const VecTypeA& a,
    const VecTypeB& b)
{
  return sqrt(arma::accu(square(a - b)));
}
```
to
```
typename VecTypeA::elem_type LMetric<2, true>::Evaluate(
    const VecTypeA& a,
    const VecTypeB& b)
{
  return arma::norm(a - b, 2);
}
```

Reply:
You're right, that might actually make this a lot easier.  I am not sure what we would need to do for other metrics, though.  If you want to change to `arma::norm(a - b, 2)` though please just do a couple simulations first to make sure that this doesn't affect the runtime.

Reply:
Turns out arma::norm doesn't allow to evaluate the norm across one of the dimensions (e.g. columns or rows), this is why it got a bit more messy. I think it would be a good idea to have this feature in Armadillo :)

Sorry about the bad code style, it was just a draft to check performance. I tried something like this:
```c++
size_t firstIndex;
  if (alreadyDidRefPoint0)
    firstIndex = 1;
  else
    firstIndex = 0;

arma::uvec newIndexes(m);

for (size_t i = 0; i < m; ++i)
  newIndexes(i) = referenceNode.Descendant(math::RandInt(firstIndex, refNumDesc));

arma::mat newRefPoints = referenceSet.cols(newIndexes);
arma::mat diff = std::move(newRefPoints).each_col() - querySet.unsafe_col(queryIndex);
arma::rowvec distances = arma::sqrt(arma::sum(arma::square(std::move(diff)), 0));

for (size_t i = 0; i < distances.n_cols; ++i)
  sample(oldSize + i) = kernel.Evaluate(distances(i));
```
after doing some benchmark, it performed worse (i.e. runtime was on average ~8% worse for this test) than the current implementation.

I think there might be better ways to implement this than the one I used (at least more elegant ones) but I'm not sure if it would be possible to get the speedup that would justify making the code a bit more messy. So I think this will stay with the current implementation :)

Anyway, let me know what you think.

Reply:
I tried using `arma::norm(a - b, 2)` for the `EuclideanDistance` and I think we got a slight speedup :)

Implemented in 5c3dc1a2fe6f397acb072330eeaf81e9f00b687a.

Reply:
Oh, I see what you mean.  That would indeed be a little messy.  There are two additional optimizations that could be done on the code you wrote, but let me write them, and then I'll write why I think it's not worth pursuing either of them:

 1) for trees that hold contiguous sets of points (the BinarySpaceTree does this), you could avoid making `newIndexes` and actually just pass a first and last index to the matrix (or perhaps even the result of a `.cols()` call to the dataset held by the tree).
 2) one could make a batch call to `kernel.Evaluate()` on the entire distances vector instead of individually calling it for each element.

However, it seems to me that the overhead of doing so way outweighs any speedup we might see, given your initial results.  We'd need to (a) create infrastructure for every tree so that we can get the entire subset of points as a matrix... or as a list of indices... or either and then use `TreeTraits<>` to figure out which to use... (b) redo all the kernels so that they have a batch `Evaluate()`... (c) debug all the metaprogramming errors that might happen... (d) hope the new API makes sense... (e) hope it even gives any speedup.

If you'd seen big speedup, I might say awesome, let me try refactoring some things and let's make it work, but in this particular case I think the effort outweighs the potential speed gain (if you agree).  Thanks for investigating this! :+1:

Reply:
I agree :) Thanks for proposing the idea.